pandas/core/series.py

```diff
@@ -2227,20 +2246,33 @@ def combine(self, other, func, fill_value=np.nan):
         Series.combine_first : Combine Series values, choosing the calling
             Series's values first
         """
+        self_is_ext = is_extension_array_dtype(self)
+        if fill_value is None:
+            if self_is_ext:
+                fill_value = self.dtype.na_value
+            else:
+                fill_value = np.nan
         if isinstance(other, Series):
             new_index = self.index.union(other.index)
             new_name = ops.get_op_result_name(self, other)
-            new_values = np.empty(len(new_index), dtype=self.dtype)
+            new_values = []
             for i, idx in enumerate(new_index):
                 lv = self.get(idx, fill_value)
                 rv = other.get(idx, fill_value)
                 with np.errstate(all='ignore'):
-                    new_values[i] = func(lv, rv)
+                    new_values.append(func(lv, rv))
         else:
             new_index = self.index
-            with np.errstate(all='ignore'):
-                new_values = func(self._values, other)
+            if not self_is_ext:
+                with np.errstate(all='ignore'):
+                    new_values = func(self._values, other)
+            else:
+                new_values = [func(lv, other) for lv in self._values]
             new_name = self.name
+
+        if (self_is_ext and self.values.is_sequence_of_dtype(new_values)):
+            new_values = self._values._from_sequence(new_values)
```

Comment:
`combine` takes a `func` as an argument. The `func` may not necessarily return objects of the `ExtensionDtype`.  For example, consider a logical operator.  So here `self.values` is the original `ExtensionArray`, but `new_values` is the result of applying the `func` element by element, and those individual results aren't necessarily of the corresponding `ExtensionDtype`.

Reply:
Thanks.

How important is it to allow coerction? The previous code certainly considered dtype-preserving functions to be the *expected* case, since the pre-allocated `new_dtype` had the same dtype as self. Only if necessary was it coerced.

Without having studied the uses of `combine`, can we say "extension arrays are not allowed to upcast in combine?". i.e. `Series[ExtensionArray].combine(Series[Any]) -> Series[ExtensionArray]`. That doesn't sound quiet right...

Anyway, my aversion is to having to perform a full scan of the data just to determine the dtype. That's what types are for :) Can we ask the user to provide an `output_dtype` argument?

Reply:
Consider the following (using the implementation in this PR):
```
In [1]: import pandas as pd
   ...: from pandas.tests.extension.decimal.array import DecimalArray, make_dat
   ...: a
   ...:

In [2]: da1= make_data()
   ...: da2= make_data()
   ...: da1[:5], da2[:5]
   ...:
Out[2]:
([Decimal('0.9100374922132099531069115982973016798496246337890625'),
  Decimal('0.559003605365540945371094494475983083248138427734375'),
  Decimal('0.31951366993722529752375294265220873057842254638671875'),
  Decimal('0.238154945978455767630066475248895585536956787109375'),
  Decimal('0.5851317119327676952167394119896925985813140869140625')],
 [Decimal('0.12525543165059660477567149428068660199642181396484375'),
  Decimal('0.68213474143447905273518472313298843801021575927734375'),
  Decimal('0.29982507800002611286771525556105189025402069091796875'),
  Decimal('0.297029189226840184545608281041495501995086669921875'),
  Decimal('0.5969224093736389402664599401759915053844451904296875')])

In [3]: s1 = pd.Series(DecimalArray(da1))
   ...: s2 = pd.Series(DecimalArray(da2))
   ...: pd.DataFrame({'s1':s1, 's2':s2}).head()
   ...:
Out[3]:
                                                  s1
                     s2
0  0.91003749221320995310691159829730167984962463...  0.12525543165059660477567149428068660199642181...
1  0.55900360536554094537109449447598308324813842...  0.68213474143447905273518472313298843801021575...
2  0.31951366993722529752375294265220873057842254...  0.29982507800002611286771525556105189025402069...
3  0.23815494597845576763006647524889558553695678...  0.29702918922684018454560828104149550199508666...
4  0.58513171193276769521673941198969259858131408...  0.59692240937363894026645994017599150538444519...

In [4]: s1.dtype, s2.dtype
Out[4]:
(<pandas.tests.extension.decimal.array.DecimalDtype at 0x15bc5ef1048>,
 <pandas.tests.extension.decimal.array.DecimalDtype at 0x15bc5ef1048>)

In [5]: cores = s1.combine(s2, lambda x1, x2: x1 if x1 < x2 else x2)
   ...: cores.head()
   ...:
Out[5]:
0    0.12525543165059660477567149428068660199642181...
1    0.55900360536554094537109449447598308324813842...
2    0.29982507800002611286771525556105189025402069...
3    0.23815494597845576763006647524889558553695678...
4    0.58513171193276769521673941198969259858131408...
dtype: decimal

In [6]: clres = s1.combine(s2, lambda x1, x2: (x1 < x2))
   ...: clres.head()
   ...:
Out[6]:
0    False
1     True
2    False
3     True
4     True
dtype: bool
```
Note that with the implementation as in this PR, we get a Series of dtype `decimal` since the results are all `Decimal` objects, and we get a `boolean` Series when the results are logical.

The previous behavior would product `object`, which is not what you want.  Given a sequence `s`, if you do `pd.Series(s)`, the internals of pandas look at the sequence `s` to determine the `dtype`. But those internals won't pick things up correctly for things that are `ExtensionDtype`.  So somehow, we need to know to call `ExtensionArray._from_sequence()` to get the result of `combine` into the right type.

The implementation is already doing this element-by-element, so we are doing a full scan of both the left and right arrays. This is an extra scan on the result.

We can add an `output_dtype` argument to `combine`, but then we'd be changing the API of `combine` and would need to decide what the default value of that parameter is.

Reply:
I would rather not do this here at all, prefering instead to dispatch to the EA itself for a ``.combine()`` method, this will be much simpler.

Reply:
Instead of introducing the `is_sequence_of_dtype` check, an alternative might be to `try: .. except` `_from_sequence()`? 

I would prefer not to add `combine` logic to the extension array

Reply:
@jorisvandenbossche My most recent commit has this change you suggested and removes `is_sequence_of_dtype`.  It required an extra check to see if it is categorical.  I also added a test which revealed a couple of other bugs.

In terms of whether `combine` should be in the EA class or in `series.py`, you guys will have to decide on that.